src/layers/basic.jl

```diff
@@ -15,35 +14,49 @@ julia> m = Chain(x -> x^2, x -> x+1);
 julia> m(5) == 26
 true
 
-julia> m = Chain(Dense(10, 5), Dense(5, 2));
+julia> m = Chain(Dense(10, 5, tanh), Dense(5, 2));
 
-julia> x = rand(10);
+julia> x = rand(10, 32);
 
 julia> m(x) == m[2](m[1](x))
 true
+
+julia> m2 = Chain(enc = Chain(Flux.flatten, Dense(10, 5, tanh)), 
+                  dec = Dense(5, 2));
+
+julia> m2(x) == (m2.dec ∘ m2.enc)(x)
+true
 ```
 """
-struct Chain{T<:Tuple}
+struct Chain{T}
   layers::T
   Chain(xs...) = new{typeof(xs)}(xs)
+  function Chain(; kw...)
+    :layers in Base.keys(kw) && throw(ArgumentError("a Chain cannot have a named layer called `layers`"))
+    isempty(kw) && return new{Tuple{}}(())
+    new{typeof(values(kw))}(values(kw))
+  end
 end
 
+Base.parent(c::Chain) = getfield(c, :layers)
+Base.propertynames(c::Chain) = (Base.keys(parent(c))..., :layers)
+Base.getproperty(c::Chain, s::Symbol) = s === :layers ? parent(c) : getproperty(parent(c), s)
+
 @forward Chain.layers Base.getindex, Base.length, Base.first, Base.last,
-  Base.iterate, Base.lastindex
+  Base.iterate, Base.lastindex, Base.keys
 
-functor(::Type{<:Chain}, c) = c.layers, ls -> Chain(ls...)
+functor(::Type{<:Chain}, c) = parent(c), ls -> Chain(ls...)
 
 applychain(::Tuple{}, x) = x
-applychain(fs::Tuple, x) = applychain(tail(fs), first(fs)(x))
+applychain(fs, x) = applychain(tail(Tuple(fs)), first(fs)(x))
 
-(c::Chain)(x) = applychain(c.layers, x)
+(c::Chain)(x) = applychain(parent(c), x)
 
-Base.getindex(c::Chain, i::AbstractArray) = Chain(c.layers[i]...)
+Base.getindex(c::Chain, i::AbstractArray) = Chain(parent(c)[i]...)
 
 function Base.show(io::IO, c::Chain)
-  print(io, "Chain(")
-  join(io, c.layers, ", ")
-  print(io, ")")
+  print(io, "Chain")
+  show(io, parent(c))
```

Comment:
This should fix the trailing comma. Is the current change here for another reason?

```suggestion
  join(io, parent(c), ", ")
```

Reply:
The change is a quick hack so that it shows you the names when there's NamedTuple. Maybe with more thought we can hack `join` & `zip` together nicely...

Reply:
Something like
```julia
_show_layers(io, layers::Tuple) = join(io, layers, ", ")
_show_layers(io, layers::NamedTuple) = join(io, map((k, v) -> "$k = $v", pairs(layers)), ", ")
```

Reply:
Oh except `ERROR: map is not defined on dictionaries`. So I made it a comprehension.